File: src/main/java/com/lcaohoanq/nocket/domain/post/Post.java
package com.lcaohoanq.nocket.domain.post;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.lcaohoanq.nocket.base.entity.BaseEntity;
import com.lcaohoanq.nocket.constant.BusinessNumber;
import com.lcaohoanq.nocket.domain.reaction.PostReaction;
import com.lcaohoanq.nocket.domain.user.User;
import com.lcaohoanq.nocket.enums.PostType;
import com.lcaohoanq.nocket.metadata.MediaMeta;
import jakarta.persistence.AttributeOverride;
import jakarta.persistence.AttributeOverrides;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Size;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "posts")
@SuperBuilder
@AttributeOverrides({
    @AttributeOverride(name = "createdAt", column = @Column(name = "post_created_at")),
    @AttributeOverride(name = "updatedAt", column = @Column(name = "post_updated_at"))
})
@EntityListeners(PostListener.class)
public class Post extends BaseEntity {

    private PostType postType;
    
    @Size(
        max = BusinessNumber.MAXIMUM_POST_CAPTION_LENGTH,
        message = """
            Caption must be less than or equal to {max} characters.
            """
    )
    private String caption;
    
    @Embedded
    private MediaMeta mediaMeta;

    @ManyToOne
    @JoinColumn(name = "user_id", referencedColumnName = "id", nullable = false)
    private User user;

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference(value = "post-reactions")
    private List<PostReaction> reactions = new ArrayList<>();

//    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
//    private List<Comment> comments = new ArrayList<>();

    // Convenience methods
    public void addReaction(PostReaction reaction) {
        reactions.add(reaction);
        reaction.setPost(this);
    }

//    public void addComment(Comment comment) {
//        comments.add(comment);
//        comment.setPost(this);
//    }
    
}

File: src/main/java/com/lcaohoanq/nocket/domain/post/Post.kt
package com.lcaohoanq.nocket.domain.post

import com.fasterxml.jackson.annotation.JsonManagedReference
import com.lcaohoanq.nocket.base.entity.BaseEntity
import com.lcaohoanq.nocket.constant.BusinessNumber
import com.lcaohoanq.nocket.constant.ValidationMessage
import com.lcaohoanq.nocket.domain.reaction.PostReaction
import com.lcaohoanq.nocket.domain.user.User
import com.lcaohoanq.nocket.enums.PostType
import com.lcaohoanq.nocket.metadata.MediaMeta
import jakarta.persistence.*
import jakarta.validation.constraints.Size

@Entity
@Table(name = "posts")
@AttributeOverrides(
    AttributeOverride(
        name = "createdAt",
        column = Column(name = "post_created_at")
    ),
    AttributeOverride(name = "updatedAt", column = Column(name = "post_updated_at"))
)
@EntityListeners(PostListener::class)
class Post : BaseEntity() {
    @Column(name = "post_type")
    private var postType: PostType? = null

    @Size(
        max = BusinessNumber.MAXIMUM_POST_CAPTION_LENGTH,
        message = ValidationMessage.CAPTION_MAX_LENGTH_MSG
    )
    private var caption: String? = null

    @Embedded
    private var mediaMeta: MediaMeta? = null

    @ManyToOne
    @JoinColumn(name = "user_id", referencedColumnName = "id", nullable = false)
    private var user: User? = null

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    @JsonManagedReference(value = "post-reactions")
    private var reactions: MutableList<PostReaction> = ArrayList()

    // Setter methods
    fun setPostType(postType: PostType) {
        this.postType = postType
    }

    fun setCaption(caption: String) {
        this.caption = caption
    }

    fun setMediaMeta(mediaMeta: MediaMeta) {
        this.mediaMeta = mediaMeta
    }

    fun setUser(user: User) {
        this.user = user
    }

    fun setReactions(reactions: MutableList<PostReaction>) {
        this.reactions = reactions
    }

    fun getUser(): User? {
        return user
    }

    // Convenience methods
    fun addReaction(reaction: PostReaction) {
        reactions.add(reaction)
        reaction.post = this
    }
}
